allennlp/modules/token_embedders/bow_token_embedder.py

```diff
@@ -0,0 +1,87 @@
+import torch
+from allennlp.modules.token_embedders.token_embedder import TokenEmbedder
+from allennlp.data import Vocabulary
+from allennlp.common import Params
+from allennlp.nn.util import get_text_field_mask
+
+
+@TokenEmbedder.register("bow_token_embedder")
+class BagOfWordsTokenEmbedder(TokenEmbedder):
+    """
+    Embeds a document in a bag of words representation. Each document is represented by
+    a vector of length vocabulary size. Each cell corresponds to a different token in
+    the vocabulary, and is populated with the number of times that token appears
+    in the document. This embedding serves as a strong baseline for document-level
+    representation learning.
+
+    Parameters
+    ----------
+    vocab_size: int
+        size of vocabulary
+    projection_dim : int, optional (default = None)
+        if specified, will project the resulting bag of words representation
+        to specified dimension.
+    """
+    def __init__(self, vocab_size: int, projection_dim: int = None) -> None:
+        super(BagOfWordsTokenEmbedder, self).__init__()
+        self.vocab_size = vocab_size
+        if projection_dim:
+            self._projection = torch.nn.Linear(vocab_size, projection_dim)
+        else:
+            self._projection = None
+        self.output_dim = projection_dim or vocab_size
+
+    def get_output_dim(self):
+        return self.num_embeddings
+
+    def compute_bow(self, tokens: torch.IntTensor) -> torch.Tensor:
+        """
+        Compute a bag of words representation with size batch_size x vocab_size
+
+        Parameters
+        ----------
+        tokens : ``Dict[str, torch.Tensor]``
+            tokens to compute bag of words representation of
+        """
+        bow_vectors = []
+        mask = get_text_field_mask({'tokens': tokens})
+        for document, doc_mask in zip(tokens, mask):
```

Comment:
Do `masked_select` and `bincount` not work on batched tensors?  Seems like it would be a lot faster if you could figure out how to make this batched.

Reply:
Yeah, I was trying to figure that out - but [`masked_select`](https://pytorch.org/docs/stable/torch.html#torch.masked_select) seems to return a 1-D tensor regardless of input tensor, and [`bincount`](https://pytorch.org/docs/stable/torch.html#torch.bincount) only takes in a 1-D tensor.